### PR TITLE
HessianRecoverer2D

### PR DESCRIPTION
**Pull request by jwallwork23**
Building upon PR #242, maybe this will be useful for other people (e.g. in #277).

To use it in practice, do something like
```
from thetis.recovery import HessianRecoverer2D

P1_ten = TensorFunctionSpace(mesh2d, 'CG', 1)
hessian_2d = Function(P1_ten)
hessian_calculator = HessianRecoverer2D(hessian_2d, solver_obj, 'elev_2d')
solver_obj.add_new_field(hessian_2d, 'hessian_2d', 'Hessian of elevation', 'Hessian2d', preproc_func=hessian_calculator.solve)
solver_obj.options.fields_to_export = ['hessian_2d']
```

### CODE REVIEW

File: thetis/recovery.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    Hessians are recoved using double :math:`L^2`
+    projection, which is implemented using a
+    mixed finite element method.
+
+    :arg hessian_2d: :class:`Function` to hold recovered Hessian.
+    :arg solver_obj: :class:`FlowSolver2d` object.
+    :kwargs: to be passed to the :class:`LinearVariationalSolver`.
+    """
+    @PETSc.Log.EventDecorator("thetis.HessianRecoverer2D.__init__")
+    def __init__(self, hessian_2d, solver_obj, field, **kwargs):
```
Comment (tkarna) [inline thetis/recovery.py:59]: I would prefer that these routines do not depend on `solver_obj` (applies to VorticityCalculator2D as well). After all, you only use it to access `solver_obj.fields[field]` so you could just pass in the field itself. This implies a little more boilerplate code by the user but you already need some, e.g. to define the TensorFunctionSpace. If this is too much we could add a helper function somewhere.

File: thetis/recovery.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            raise ValueError(f'Dimension {dim} not supported')
+        if element_continuity(fs.ufl_element()).horizontal != 'cg':
+            raise NotImplementedError
+        if fs.dof_dset.dim != (2, 2):
+            raise ValueError('Expecting a square tensor function')
+        n = FacetNormal(mesh)
+
+        # Setup function spaces
+        Sigma = hessian_2d.function_space()
+        V = solver_obj.function_spaces.P1v_2d
```
Comment (tkarna) [inline thetis/recovery.py:77]: You could call `thetis.get_functionspace` instead

**Review by tkarna — COMMENTED** [review]
Thanks, this is great! I wonder whether "recovery" is the right term to use here? These two routines calculate some diagnostic fields, so the py file could be "diagnostics", "diagnostic_fields", or "postprocessing" or such? Gradient and Hessian could go under gradient estimation umbrella, but e.g. vorticity maybe not.

**Comment by jwallwork23** [comment]
> Thanks, this is great! I wonder whether "recovery" is the right term to use here? These two routines calculate some diagnostic fields, so the py file could be "diagnostics", "diagnostic_fields", or "postprocessing" or such? Gradient and Hessian could go under gradient estimation umbrella, but e.g. vorticity maybe not.

Good point. The vorticity field involves derivatives to be recovered, but I agree "diagnostics" would be a better name.

File: thetis/recovery.py (hunk 1 above)
Comment (jwallwork23) [reply thetis/recovery.py:59]: Sure, I can do that.

**Comment by jwallwork23** [comment]
I reworked both ``VorticityCalculator2D`` and ``HessianRecoverer2D`` so that they do not depend on the solver object. I also added them to tests, to check that they work and provide example usage.

Here is the magnitude of the elevation Hessian for ``swe2d/test_anisotropic.py``
hessian_elev

**Comment by tkarna** [comment]
Should we also include tests against analytical solutions to check that the vorticity/hessian are computed correctly? Could go under `test/operations/`.

**Comment by jwallwork23** [comment]
> Should we also include tests against analytical solutions to check that the vorticity/hessian are computed correctly? Could go under `test/operations/`.

Sure, I'll take a look at that later in the week.

**Review by tkarna — APPROVED** [review]
Thanks for the updates, looks good to me!

**Pull request merged**